services/elasticache/pkg/resource/replication_group/custom_update_api.go

```diff
@@ -110,144 +180,184 @@ func (rm *resourceManager) updateReplicaCount(
+func (rm *resourceManager) newDecreaseReplicaCountRequestPayload(
+	desired *resource,
+) (*svcsdk.DecreaseReplicaCountInput, error) {
+	res := &svcsdk.DecreaseReplicaCountInput{}
+	desiredSpec := desired.ko.Spec
 
 	res.SetApplyImmediately(true)
-	if r.ko.Spec.ReplicationGroupID != nil {
-		res.SetReplicationGroupId(*r.ko.Spec.ReplicationGroupID)
+	if desiredSpec.ReplicationGroupID != nil {
+		res.SetReplicationGroupId(*desiredSpec.ReplicationGroupID)
 	}
-	if r.ko.Spec.NumNodeGroups != nil {
-		res.SetNewReplicaCount(*r.ko.Spec.ReplicasPerNodeGroup)
+	if desiredSpec.ReplicasPerNodeGroup != nil {
+		res.SetNewReplicaCount(*desiredSpec.ReplicasPerNodeGroup)
 	}
 
-	if r.ko.Spec.NodeGroupConfiguration != nil {
-		f13 := []*svcsdk.ConfigureShard{}
-		for _, f13iter := range r.ko.Spec.NodeGroupConfiguration {
-			f13elem := &svcsdk.ConfigureShard{}
-			if f13iter.NodeGroupID != nil {
-				f13elem.SetNodeGroupId(*f13iter.NodeGroupID)
+	if desiredSpec.NodeGroupConfiguration != nil {
+		shardsConfig := []*svcsdk.ConfigureShard{}
+		for _, desiredShard := range desiredSpec.NodeGroupConfiguration {
+			shardConfig := &svcsdk.ConfigureShard{}
+			if desiredShard.NodeGroupID != nil {
+				shardConfig.SetNodeGroupId(*desiredShard.NodeGroupID)
 			}
-			if f13iter.ReplicaCount != nil {
-				f13elem.SetNewReplicaCount(*f13iter.ReplicaCount)
+			if desiredShard.ReplicaCount != nil {
+				shardConfig.SetNewReplicaCount(*desiredShard.ReplicaCount)
 			}
-			f13elemf2 := []*string{}
-			if f13iter.PrimaryAvailabilityZone != nil {
-				f13elemf2 = append(f13elemf2, &(*f13iter.PrimaryAvailabilityZone))
+			shardAZs := []*string{}
+			if desiredShard.PrimaryAvailabilityZone != nil {
+				shardAZs = append(shardAZs, desiredShard.PrimaryAvailabilityZone)
 			}
-			if f13iter.ReplicaAvailabilityZones != nil {
-				for _, f13elemf2iter := range f13iter.ReplicaAvailabilityZones {
-					var f13elemf2elem string
-					f13elemf2elem = *f13elemf2iter
-					f13elemf2 = append(f13elemf2, &f13elemf2elem)
+			if desiredShard.ReplicaAvailabilityZones != nil {
+				for _, desiredAZ := range desiredShard.ReplicaAvailabilityZones {
+					shardAZs = append(shardAZs, desiredAZ)
 				}
-				f13elem.SetPreferredAvailabilityZones(f13elemf2)
 			}
-			f13 = append(f13, f13elem)
+			if len(shardAZs) > 0 {
+				shardConfig.SetPreferredAvailabilityZones(shardAZs)
+			}
+			shardsConfig = append(shardsConfig, shardConfig)
 		}
-		res.SetReplicaConfiguration(f13)
+		res.SetReplicaConfiguration(shardsConfig)
 	}
 
 	return res, nil
 }
 
-// new(DecreaseReplicaCount)RequestPayload returns an SDK-specific struct for the HTTP request
-// payload of the Create API call for the resource
-func (rm *resourceManager) newDecreaseReplicaCountRequestPayload(
-	r *resource,
-) (*svcsdk.DecreaseReplicaCountInput, error) {
-	res := &svcsdk.DecreaseReplicaCountInput{}
+// newUpdate(ShardConfiguration)RequestPayload returns an SDK-specific struct for the HTTP request
```

Comment:
Can you remove the parens around ShardConfiguration above?

Reply:
Updated source file.